### PR TITLE
Setup basic functionality

### PR DESCRIPTION
**Pull request by seancorc**
## Overview

- Add routes for creating, deleting, updating the expiration date, and getting the active announcements, if any (An announcement is active if its expiration date is after the current date). 
- Add basic announcement table schema.

## Test Coverage

I tested each route using Postman.

## Next Steps
1. Implement the ability to specify which iOS/Android apps the announcement should be presented on (e.g. not announcing eatery specific info on transit). 
2. Create an API doc for iOS/Android devs.

</details>

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -0,0 +1,100 @@
+from flask import Flask, request
+import json
+from db import Announcement, db
+from datetime import datetime
+
+app = Flask(__name__)
+
+db_filename = "annoucement.db"
```
Comment (manishsaha) [inline app.py]: should be announcements.db

File: app.py [hunk 2]
```diff
@@ -0,0 +1,100 @@
+from flask import Flask, request
+import json
+from db import Announcement, db
+from datetime import datetime
```
Comment (manishsaha) [inline app.py]: alphabetize

File: app.py [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+INVALID_DATE_ERROR = json.dumps(
+    {"success": False, "error": "Please input a valid date of the form 'y/m/d'"}
+)
+INVALID_ANNOUNCEMENT_ERROR = json.dumps(
+    {"success": False, "error": "No announcement with that name exists"}
+)
+INVALID_NAME_ERROR = json.dumps(
+    {"success": False, "error": "Announcement with that name already exists"}
+)
+SUCCESFUL_RESPONSE = json.dumps({"success": True})
```
Comment (manishsaha) [inline app.py]: this should probably go in a constants file

File: app.py [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+
+@app.route("/create/", methods=["POST"])
+def create_announcment():
```
Comment (manishsaha) [inline app.py]: announcement*

File: app.py (hunk 3 above)
Comment (manishsaha) [inline app.py]: SUCCESSFUL*

File: app.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        ctaAction=ctaAction,
+        ctaText=ctaText,
+        expirationDate=expiration,
+        imageUrl=imageUrl,
+        subject=subject,
+        name=name,
+    )
+    db.session.add(announcement)
+    db.session.commit()
+    return SUCCESFUL_RESPONSE, 200
```
Comment (manishsaha) [inline app.py]: SUCCESSFUL*

File: app.py [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+
+@app.route("/update_expiration_date/", methods=["PUT"])
```
Comment (manishsaha) [inline app.py]: I'm not sure if we use any PUT methods in any of our endpoints, maybe this should just be a POST? would confirm with others tho

File: app.py [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+    try:
+        expiration = datetime.strptime(expirationDate, "%Y/%m/%d")
+    except:
+        return INVALID_DATE_ERROR, 400
+    announcement = Announcement.query.filter_by(name=name).first()
+    if announcement is None:
+        return INVALID_ANNOUNCEMENT_ERROR, 400
+    announcement.expirationDate = expiration
+    db.session.commit()
+    return SUCCESFUL_RESPONSE, 200
```
Comment (manishsaha) [inline app.py]: typo, as above

File: app.py [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+
+
+@app.route("/delete/<name>/", methods=["DELETE"])
+def delete_announcement(name):
+    announcment = Announcement.query.filter_by(name=name).first()
+    if announcment is None:
+        return INVALID_ANNOUNCEMENT_ERROR, 400
+    db.session.delete(announcment)
```
Comment (manishsaha) [inline app.py]: announcement*

File: app.py [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+    db.session.commit()
+    return SUCCESFUL_RESPONSE, 200
```
Comment (manishsaha) [inline app.py]: typo here too

File: app.py [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+
+@app.route("/get_announcements/", methods=["GET"])
+def get_announcements():
+    active_announcements = Announcement.query.filter(
+        Announcement.expirationDate > datetime.now()
+    ).all()
+    if active_announcements == []:
```
Comment (manishsaha) [inline app.py]: you can probably just check `if not active_announcements`

File: app.py [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+        return json.dumps({"data": None}), 200
+    res = {"data": [announcement.serialize() for announcement in active_announcements]}
+    return json.dumps(res), 200
+
+
+if __name__ == "__main__":
+    app.run(host="0.0.0.0", port=5000, debug=True)
```
Comment (manishsaha) [inline app.py]: def create environment variables for port, filenames, etc

**Review by manishsaha — CHANGES_REQUESTED** [review]

File: app.py (hunk 9 above)
Comment (seancorc) [reply app.py]: Thanks for feedback, I really need to learn how to spell lol.

File: app.py (hunk 6 above)
Comment (seancorc) [reply app.py]: I think this is actually supposed to be PATCH as "The PATCH method requests that a set of changes described in the request entity be applied to the resource identified by the Request- URI." (https://tools.ietf.org/html/rfc5789). 

It should work as intended either way so fine to change it if desired.

**Review by meganle — COMMENTED** [review]
Can we also add the black linting with the precommit hook? @sahamanish99 can help you with this!

File: app.py [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+@app.route("/create/", methods=["POST"])
+def create_announcment():
+    post_body = json.loads(request.data)
+    body = post_body.get("body")
+    ctaAction = post_body.get("ctaAction")
+    ctaText = post_body.get("ctaText")
+    expirationDate = post_body.get("expirationDate")
+    imageUrl = post_body.get("imageUrl")
+    name = post_body.get("name")
+    subject = post_body.get("subject")
```
Comment (kq-li) [inline app.py]: should add a way to fail gracefully if any of these fields are not included

File: app.py (hunk 6 above)
Comment (kq-li) [reply app.py]: prefer calling this route `/update/` and having a way to update the other fields as well

File: app.py [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+    )
+    db.session.add(announcement)
+    db.session.commit()
+    return SUCCESFUL_RESPONSE, 200
+
+
+@app.route("/update_expiration_date/", methods=["PUT"])
+def update_expiration_date():
+    post_body = json.loads(request.data)
+    name = post_body.get("name")
```
Comment (kq-li) [inline app.py]: incorporate ID here instead of name (and maybe make it a route parameter)

File: app.py [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+def create_announcment():
+    post_body = json.loads(request.data)
+    body = post_body.get("body")
+    ctaAction = post_body.get("ctaAction")
+    ctaText = post_body.get("ctaText")
+    expirationDate = post_body.get("expirationDate")
+    imageUrl = post_body.get("imageUrl")
+    name = post_body.get("name")
+    subject = post_body.get("subject")
+    # Name is an identifier, so there should not be any duplicates.
```
Comment (kq-li) [inline app.py]: i think it might be better to actually generate an ID and not treat name as a primary key here

File: app.py [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+        return INVALID_DATE_ERROR, 400
+    announcement = Announcement.query.filter_by(name=name).first()
+    if announcement is None:
+        return INVALID_ANNOUNCEMENT_ERROR, 400
+    announcement.expirationDate = expiration
+    db.session.commit()
+    return SUCCESFUL_RESPONSE, 200
+
+
+@app.route("/delete/<name>/", methods=["DELETE"])
```
Comment (kq-li) [inline app.py]: ID again

File: app.py [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+def delete_announcement(name):
+    announcment = Announcement.query.filter_by(name=name).first()
+    if announcment is None:
+        return INVALID_ANNOUNCEMENT_ERROR, 400
+    db.session.delete(announcment)
+    db.session.commit()
+    return SUCCESFUL_RESPONSE, 200
+
+
+@app.route("/get_announcements/", methods=["GET"])
```
Comment (kq-li) [inline app.py]: i prefer `/active/`, and you don't need to include the methods if it's just GET

File: app.py [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+def get_announcements():
+    active_announcements = Announcement.query.filter(
+        Announcement.expirationDate > datetime.now()
```
Comment (kq-li) [inline app.py]: technically there should be a trailing comma here lol

File: app.py [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+    ).all()
+    if active_announcements == []:
+        return json.dumps({"data": None}), 200
```
Comment (kq-li) [inline app.py]: missing success field

File: app.py (hunk 11 above)
Comment (kq-li) [inline app.py]: turn off debug mode

**Review by kq-li — CHANGES_REQUESTED** [review]

**Comment by young-k** [comment]
@seancorc can you mark comments as resolved once they are?

File: app.py (hunk 17 above)
Comment (seancorc) [reply app.py]: Black is deleting it when I try haha.

File: app.py [hunk 19]
```diff
@@ -0,0 +1,106 @@
...
+    return Response.SUCCESSFUL_RESPONSE, 200
+
+
+@app.route("/delete/<id>/", methods=["DELETE"])
+def delete_announcement(id):
+    announcment = Announcement.query.get(id)
+    if announcment is None:
+        return Response.INVALID_ANNOUNCEMENT_ERROR, 400
+    db.session.delete(announcment)
+    db.session.commit()
```
Comment (manishsaha) [inline app.py:85]: announcement* in lines 81, 82 and 84

File: app.py [hunk 20]
```diff
@@ -0,0 +1,106 @@
...
+
+@app.route("/active/")
+def get_announcements():
+    active_announcements = (
+        Announcement.query.filter(Announcement.expirationDate > datetime.now())
+        .filter(Announcement.startDate < datetime.now())
+        .all()
+    )
+    if not active_announcements:
+        return json.dumps({"success": True, "data": None}), 200
```
Comment (manishsaha) [inline app.py]: this should just be `“data”: []` for uniformity

Comment (manishsaha) [reply app.py]: and then i think u won’t even need this conditional tbh

**Review by manishsaha — CHANGES_REQUESTED** [review]
thanks for implementing all the changes, just a few more things!

**Review by manishsaha — APPROVED** [review]
amazing

File: app.py [hunk 21]
```diff
@@ -0,0 +1,104 @@
...
+    )
+    res = {
+        "success": True,
+        "data": [announcement.serialize() for announcement in active_announcements],
+    }
+    return json.dumps(res), 200
+
+
+if __name__ == "__main__":
+    app.run(host="0.0.0.0", port=environ["PORT"], debug=False)
```
Comment (kq-li) [inline app.py]: you can just leave out the debug here

File: db.py [hunk 22]
```diff
@@ -0,0 +1,27 @@
+from flask_sqlalchemy import SQLAlchemy
+
+db = SQLAlchemy()
+
+
+class Announcement(db.Model):
+    __table_name = "announcement"
```
Comment (kq-li) [inline db.py]: prefer "announcements"

**Review by kq-li — APPROVED** [review]
sick

**Pull request merged**